File: go/models/query_result.go
package models

// CargoEvent represents a single cargo history event.
type CargoEvent struct {
	UnitName   string `xml:"unitName"`
	EventName  string `xml:"eventName"`
	ReasonName string `xml:"reasonName"`
	EventDate  string `xml:"eventDate"`
	EventTime  string `xml:"eventTime"`
	CityName   string `xml:"cityName"`
	TownName   string `xml:"townName"`
}

// QueryItemDetail represents the detailed item information for a query.
type QueryItemDetail struct {
	TrackingUrl       string       `xml:"trackingUrl"`
	ReceiverCustName  string       `xml:"receiverCustName"`
	DepartureUnitName string       `xml:"departureUnitName"`
	DeliveryDate      string       `xml:"deliveryDate"`
	DeliveryTime      string       `xml:"deliveryTime"`
	CargoHistory      []CargoEvent `xml:"invDocCargoDAMVOArray>InvDocCargoDAMVO"`
}

// QueryDetail represents the result detail for a single query.
type QueryDetail struct {
	CargoKey         string           `xml:"cargoKey"`
	InvoiceKey       string           `xml:"invoiceKey"`
	OperationStatus  string           `xml:"operationStatus"`
	OperationMessage string           `xml:"operationMessage"`
	ErrCode          int              `xml:"errCode"`
	ErrMessage       string           `xml:"errMessage"`
	ItemDetail       *QueryItemDetail `xml:"invDocCargoVO"`
}

// IsSuccess returns true if this detail has no error.
func (d *QueryDetail) IsSuccess() bool {
	return d.ErrCode == 0
}

// QueryResult represents the response from queryShipment.
type QueryResult struct {
	OutFlag      int           `xml:"outFlag"`
	OutResult    string        `xml:"outResult"`
	SenderCustId int           `xml:"senderCustId"`
	Count        int           `xml:"count"`
	Details      []QueryDetail `xml:"shippingDeliveryDetailVO"`
}

// IsSuccess returns true if the operation was successful.
func (r *QueryResult) IsSuccess() bool {
	return r.OutFlag == 0
}

File: dotnet/Models/QueryResult.cs
namespace YurticiKargo.Models;

public class CargoHistoryEvent
{
    public string UnitName { get; set; } = string.Empty;
    public string EventName { get; set; } = string.Empty;
    public string ReasonName { get; set; } = string.Empty;
    public string EventDate { get; set; } = string.Empty;
    public string EventTime { get; set; } = string.Empty;
    public string CityName { get; set; } = string.Empty;
    public string TownName { get; set; } = string.Empty;
}

public class ItemDetail
{
    public string TrackingUrl { get; set; } = string.Empty;
    public string ReceiverCustName { get; set; } = string.Empty;
    public string DepartureUnitName { get; set; } = string.Empty;
    public string DeliveryDate { get; set; } = string.Empty;
    public string DeliveryTime { get; set; } = string.Empty;
    public List<CargoHistoryEvent> CargoHistory { get; set; } = new();
}

public class QueryDetail
{
    public string CargoKey { get; set; } = string.Empty;
    public string InvoiceKey { get; set; } = string.Empty;
    public string OperationStatus { get; set; } = string.Empty;
    public int OperationCode { get; set; }
    public string OperationMessage { get; set; } = string.Empty;
    public int ErrCode { get; set; }
    public string ErrMessage { get; set; } = string.Empty;
    public ItemDetail? ItemDetail { get; set; }

    public bool IsSuccess() => ErrCode == 0;
}

public class QueryResult
{
    public int OutFlag { get; set; }
    public string OutResult { get; set; } = string.Empty;
    public long SenderCustId { get; set; }
    public int Count { get; set; }
    public List<QueryDetail> Details { get; set; } = new();

    public bool IsSuccess() => OutFlag == 0;
}
